Obviates Forks of Gem Dependencies

DanielWright opened the pull request:
Using custom forks of gems introduces a lot of maintainability headaches down the road. The juice isn't worth the squeeze: it's simple enough to include the AM and Solidus gateway implementations directly within our gem.

This PR proposes to do so, as well as consolidating all gem dependencies in the gemspec.

One complication: @cabouffard's AM gateway was test-driven with Test::Unit. The unit and integration tests will need to be ported to RSpec before this PR can be considered for merge.
- [x] Remove dependencies on forked gems
- [x] Consolidate all gem dependencies in the gemspec
- [x] Integrate ActiveMerchant and Solidus gateways
- [x] Maintain test coverage

Commit:
Preliminary Cleanup of Gemfile

It's strongly preferable to keep dependencies in the gemspec, rather than the
Gemfile, particularly for base dependencies. This allows RubyGems/Bundler to
resolve the dependency tree. The exception is, for obvious reasons, gems not
hosted on RubyGems.

I also removed a development dependency on 'braintree', on which the test suite
doesn't appear to depend. (Maybe it did at one point?)

Commit:
Eliminates Use of Forked Gems

Using custom forks of gems introduces a lot of maintainability headaches down
the road. The juice isn't worth the squeeze: it's simple enough to include the
AM and Solidus gateway implementations directly within our gem.

One complication: @cabouffard's AM gateway was test-driven with Test::Unit. The
unit and integration tests will need to be ported to RSpec.

Commit:
First Attempt at Porting Shopify AM Gateway Unit Tests

There's plenty of opportunity for improvement here. I'm not sure how much I
will contribute, though, as I'm not sure this is a unit test. To be discussed.

DanielWright commented:
I've now ported the ActiveMerchant Shopify gateway's unit tests over into RSpec. They're a bit of a mess. I would clean them up some more, but I think they (and the gateway itself) bear some discussion.

There are two service objects defined in the same file as the gateway: `ShopifyRefunder` and `ShopifyVoider`. They've been extracted into the global namespace, but the unit tests treat them as private classes. I really like the notion of treating `ActiveMerchant::Billing::ShopifyGateway` as a real [gateway object](http://martinfowler.com/eaaCatalog/gateway.html), but the execution has been a bit haphazard. If they are properly extracted, the unit tests should only set expectations on delegating the work of `#refund` and `#void` to those objects (and the object dependencies should be injected, if possible). Then those objects should be individually unit-tested. If they're going to live on only as private classes that organize the implementation, then references to them shouldn't appear in the unit tests. In either case, the unit tests will need more than just tidying up, which is outside the scope of this PR.

I've also opted _not_ to port over the [end-to-end tests](https://github.com/DynamoMTL/active_merchant_new/blob/feature/shopify/prototype-implementation/test/unit/gateways/remote_shopify_test.rb), because I think they essentially duplicate the existing refund request specs. I'm open to disagreement on this.

thisiscab commented:
@DanielWright 

I do not understand how you're expecting the code structure to be after your second paragraph.

I've follow the convention of the ActiveMerchant gem and implemented the end-to-end tests like all the other Gateway over here https://github.com/DynamoMTL/active_merchant_new/tree/feature/shopify/prototype-implementation/test/remote/gateways

We have to be on the same wavelength when it comes to implementation details and what has to be done. If you opted not to port the test and I have opted to write them, then we are losing valuable time.

thisiscab committed:
WIP - Add unit tests to the shopify refunder

thisiscab committed:
Finish up the tests for the shopify refunder service

thisiscab committed:
Add specs for the shopify voider service

thisiscab commented:
@DanielWright 
What do you think of those commits?

Are we ready to merge?

DanielWright commented on Gemfile:
```diff
@@ -1,3 +1,7 @@
 
 gemspec
+
+group :test do
```
This should be superfluous given https://github.com/glossier/solidus_retail/pull/4/files#diff-134acb9c8b297af74c81f7263868fad6R27

thisiscab replied:
`binding.pry` is not accessible when running specs if not declared in the `test` group. This is why I have added it there.

DanielWright replied:
That's because Rails will automatically `require` things in the Gemfile, but not things from the gemspec. If you `require 'pry'; binding.pry` that should work.

thisiscab replied:
It's more of a convenience to have it there, so we don't have to require it only when we want it. But I see the value of requiring it when we want it so I'll remove it from the gem and see how it goes! 

I know I could put the require at the top of the spec_helper file but at the same time I've seem numero project that juste includes `pry` when they want to use it. I'll use that opportunity to give it a try.

braidn replied:
`require pry` could also be added to the spec_helper, giving us access to `binding.pry` without the require correct?

thisiscab replied:
Totally, I am unsure how it gets loaded though. 

Would it always be loaded even though we are never using pry? Would it slow down the spec suite?

DanielWright commented on spec/services/shopify_voider_spec.rb:
```diff
@@ -0,0 +1,45 @@
+
+  context '.initialize' do
+    it 'successfully does it\'s thing' do
+      expect(subject).to be_truthy
```
What value does this expectation provide (other than the test-coverage number 😄 )?

thisiscab replied:
Sometimes I write stupid stuff in my initializer and it just crashes because I call wrong stuff. 

This makes sure that the initializer actually works with the expected parameters and that it doesn't crash.

thisiscab replied:
I would agree that this would get surfaced in other tests, but most of the time I write that spec then I write my initializer so it helps me validate that it works.

braidn replied:
I also write these style of tests. I think they are super handy when doing TDD but ultimately could be deleted as new, much more meaningful tests appear. Never be afraid to delete tests.

DanielWright commented on spec/services/shopify_voider_spec.rb:
```diff
@@ -0,0 +1,45 @@
+    end
+
+    it 'calls the shopify refunder' do
+      expect(refunder_instance).to receive(:perform)
```
Typically, when unit-testing a "command" method, you want to set an expectation that the message sent to the external object is correctly formed, ie. that you're sending the correct arguments. In the implementation, you do this during instantiation. You'll probably want to add expectations about how what arguments the external object is expected to receive in the messages you send to it. For example, the voider transforms the transaction's amount by multiplying it by 100.

thisiscab replied:
I am unsure about that comment, I'd like to dicuss to either of you, if you ever have the chance. 

I wrote a big paragraph but I had trouble zoning on what I was trying to explain.

DanielWright commented on spec/services/shopify_voider_spec.rb:
```diff
@@ -0,0 +1,45 @@
+  subject { described_class.new(transaction_id, order_id) }
+
+  context '.initialize' do
+    it 'successfully does it\'s thing' do
```
Linting note: you can probably go ahead and use double-quotes to avoid unsightly escaping of characters. 😉

braidn replied:
This makes me 'twitch a little as well'. It also could be written as `it is` and still be fine.

DanielWright commented on spec/services/shopify_voider_spec.rb:
```diff
@@ -0,0 +1,45 @@
+
+require_relative '../../lib/active_merchant/billing/gateways/shopify.rb'
+
+describe ShopifyVoider do
```
Could we establish a namespace module to put things like this in? `Spree::Retail` maybe?

DanielWright commented on spec/services/shopify_voider_spec.rb:
```diff
@@ -0,0 +1,45 @@
+
+  context '.perform' do
+    before do
+      allow(ShopifyRefunder).to receive(:new).and_return(refunder_instance)
```
What are your feelings about injecting dependencies like `ShopifyRefunder` and `ShopifyAPI::Transaction` as a matter of style?

DanielWright commented on spec/services/shopify_refunder_spec.rb:
```diff
@@ -0,0 +1,109 @@
+      subject { described_class.new(credited_money_in_cents, transaction_id, options) }
+
+      it 'performs a refund on shopify' do
+        expect(ShopifyAPI::Refund).to receive(:create)
```
Similar to another note I've left, my rule of thumb for `expect` vs. `allow`:
- `allow(something).to receive(:message).and_return(stub)`
- `expect(something).to receive(:message).with(arguments)`

It's pretty important, for documentation and for robustness, to set explicit expectations arguments when exercising a method-call, but it's generally overkill for stubs.

thisiscab replied:
```
        refund = refunder_interface.create(order_id: order_id,
                                           shipping: { amount: 0 },
                                           note: refund_reason,
                                           notify: false,
                                           restock: false,
                                           transactions: [{
                                             parent_id: transaction.id,
                                             amount: amount_to_dollars(credited_money),
                                             gateway: 'shopify-payments',
                                             kind: 'refund'
                                           }])
```

That's a lot of argument, you would recommend having all of them written in the spec itself?

thisiscab replied:
`expect(refunder_interface).to receive(:create).with(hash_including(:order_id, :shipping, :note, :notify, :restock, :transactions))`

thisiscab replied:
Would it be better to write a custom matcher or have a private method in the spec that would encapsulate the required hash?

DanielWright replied:
I would include all the arguments. One of the cool things about unit tests like this is they reflect the complexity of the object under test. If your tests are complicated and difficult to maintain, it probably means _your object is too_. It's a 🚨 that maybe your object needs some refactoring. In this case, maybe your refunder shouldn't be responsible for building all those arguments? Is there an opportunity to decompose that responsibility into smaller SRP-friendly objects?

DanielWright replied:
Consider all the arguments to `create`. To provide the most thorough coverage, you should write unit-tests covering every permutation. That'd be _bonkers_, right?

thisiscab replied:
Yes! I'll see what I can do!

thisiscab replied:
What do you think of something of that kind? @DanielWright 
https://github.com/glossier/solidus_retail/commit/35e9031aabbd163ad67bdd957a52eb09aa23426d

It's only a prototype, it's not correctly implemented but it gives a general idea on how I would see stuff. 
Since the arguments and the api are tightly linked together, having a class that would tie both seemed to be like a viable idea.

DanielWright commented on spec/services/shopify_refunder_spec.rb:
```diff
@@ -0,0 +1,109 @@
+        expect(result).to be_a(ActiveMerchant::Billing::Response)
+      end
+
+      it 'returns an ActiveMerchant succesfull response' do
```
Linting note: "successful" not "succesfull"

DanielWright commented on spec/services/shopify_refunder_spec.rb:
```diff
@@ -0,0 +1,109 @@
+
+      it 'returns an ActiveMerchant succesfull response' do
+        result = subject.perform
+        expect(result.success?).to be_truthy
```
How do you feel about RSpec's magical predicate matchers? In this case, you could do `expect(result).to be_success`. RSpec chops the `be_` off, appends a `?`, and uses that for the expectation.

thisiscab replied:
TIL

DanielWright commented on spec/services/shopify_refunder_spec.rb:
```diff
@@ -0,0 +1,109 @@
+
+      it 'returns an ActiveMerchant response' do
+        result = subject.perform
+        expect(result).to be_a(ActiveMerchant::Billing::Response)
```
Shouldn't it be the responsibility of the ActiveMerchant gateway to build an ActiveMerchant response?

DanielWright commented on spec/services/shopify_refunder_spec.rb:
```diff
@@ -0,0 +1,109 @@
+
+      it 'throws an error' do
+        cause = ->{ subject.perform }
+        expect(&cause).to raise_error(ActiveMerchant::Billing::ShopifyGateway::CreditedAmountBiggerThanTransaction)
```
Similarly, wouldn't this object raise its own errors rather than borrowing ActiveMerchant's?

thisiscab replied:
Totally, but then both `refunder` and `voider` have the possibility to return the same error for example `TransactionNotFoundError`. I thought that defining that error in the parent class would reduce code duplication. 

Do you have a suggestion of where else I should define that error if not on the parent class? Is it a good idea to define the same error individually in each classes?

DanielWright replied:
I would suggest either a common namespace module (e.g. `Spree::Retail::Shopify`) or a superclass from which both inherit. The first option makes more sense to me, because I don't think inheritance makes sense. What do you think?

DanielWright commented on spec/services/shopify_refunder_spec.rb:
```diff
@@ -0,0 +1,109 @@
+
+    context 'when the refunds contains an error' do
+      let(:pos_refund) { double('refund', errors: errors) }
+      let(:errors) { double('errors', messages: ['I am ERROR']) }
```
NO I AM ERROR

DanielWright commented on spec/services/shopify_refunder_spec.rb:
```diff
@@ -0,0 +1,109 @@
+  end
+
+  context '.initialize' do
+    subject { described_class.new(credited_money_in_cents, transaction_id, options) }
```
I'm a fan lately of keyword arguments to remove the cognitive overhead of argument ordering, and more explicit documentation of the interface. What are your thoughts on that notion?

thisiscab replied:
Good idea, let's try that :)

thisiscab replied:
What if one of the required parameters is not entered, we raise an error?

For example, the credited_money_in_cents is a required parameter in the method above. If the user is allowed to enter whatever params they want, then we need to check the required parameters and raise an error if they are not filled? I know how to achieve that, but is there an example of something you have implemented that follows that behavior?

DanielWright commented on spec/lib/active_merchant/billing/gateways/shopify_spec.rb:
```diff
@@ -0,0 +1,107 @@
+      end
+    end
+
+    describe "#refund" do
```
Can this be dramatically pared down because these behaviours have been delegated to ShopifyRefunder?

DanielWright commented on spec/lib/active_merchant/billing/gateways/shopify_spec.rb:
```diff
@@ -0,0 +1,107 @@
+      end
+    end
+
+    describe "#void" do
```
Can this be dramatically pared down because these behaviours have been delegated to `ShopifyVoider`?

thisiscab committed:
Remove pry from test environment

thisiscab committed:
Namespace the service classes

thisiscab committed:
Dependency inject the required classes

This makes it easier to override if need be and it facilitates testing
those services.

thisiscab committed:
Fix typo

thisiscab committed:
Better use of the RSpec matchers

Instead of checking the trueness of the value, I simply use the
be_success matcher.

thisiscab committed:
Allow the ActiveMerchant Gateway to throw it's own response

Instead of having the voider or refunder throwing an
ActiveMerchant::Billing::Response, the actual Gateway will throw that
reponse depending on the response type received by the voider or
refunder.

thisiscab committed:
Add proper namespacing for general errors and properties

thisiscab committed:
Correctly namespace files / method calls

thisiscab committed:
Refactor the specs + make everything green

thisiscab commented:
Alright! Pretty much everything is refactored with what you guys have suggested.

Things that are left to do:
1. Figure out a better way to pass the long list of arguments to the ShopifyAPI
2. QUOTE: "Typically, when unit-testing a "command" method, you want to set an expectation that the message sent to the external object is correctly formed, ie. that you're sending the correct arguments. In the implementation, you do this during instantiation. You'll probably want to add expectations about how what arguments the external object is expected to receive in the messages you send to it. For example, the voider transforms the transaction's amount by multiplying it by 100."
3. Pass the parameters hash a hash instead of having to specify them in a certain order.
   - I have written something real quick, what do you guys think of that: https://github.com/glossier/solidus_retail/commit/35e9031aabbd163ad67bdd957a52eb09aa23426d

I'll continue working on it then ping both of you to make a second pass!

braidn commented:
Beginning the ole read through here.

braidn commented on spec/services/shopify_refunder_spec.rb:
```diff
@@ -1,53 +1,108 @@
+      subject { described_class.new(credited_money_in_cents, transaction_id, options) }
+
       it 'throws an error' do
+        cause = ->{ subject.perform }
```
Does this read easier to you Cab? Could we use `expect { subject.perform }.to raise_error(...)` and make it a magic one liner without the use of a staby lambda ?

thisiscab replied:
This is how it was written previously but they were changed to that, so I kept it like that.

braidn commented on app/services/spree/retail/shopify_refunder.rb:
```diff
@@ -0,0 +1,63 @@
+module Spree
+  module Retail
+    class ShopifyRefunder
```
AMAZE!

thisiscab committed:
Add spec for the refund decorator

thisiscab committed:
Use module convention for specs

braidn commented on spec/services/shopify_voider_spec.rb:
```diff
@@ -6,14 +6,12 @@
   let(:refunder_instance) { double('refunder_instance', perform: true) }
+  let(:refunder_class) { double('refunder_class', new: refunder_instance) }
 
-  before do
```
I think this diff is a great example of how dependency injection can clean up tests.

braidn commented on spec/services/shopify_refunder_spec.rb:
```diff
@@ -46,7 +46,7 @@
       end
 
-      it 'returns an ActiveMerchant succesfull response' do
+      it 'returns an ActiveMerchant successful response' do
```
it's super hard since successfully is written with two 'L's. BAH Engrish!

braidn commented on app/model/spree/refund_decorator.rb:
```diff
@@ -1,6 +1,10 @@
+module Spree
+  module RefundDecorator
+    def pos_order_id
+      # NOTE(cab): Should we use delegates here?
```
I am not apposed to comments as long as they bring value to the table. I am not too sure what type of value this provides except maybe leading the next dev down a potential refactor hole. I would also love it if we kept naming (even if it was a nickname) out of the comment. `git blame` or the nicer (aliased) `git source` should be good enough to drudge up who the note was created by.

thisiscab replied:
The note written there was mainly used for code-reviewer to give me feedback. It's an open question :)

Commit:
Resolves Two Spec-Order Dependencies

1. Run in isolation, the Spree::Gateway::ShopifyGateway spec will fail because
   the Spree::Gateway class will not have been loaded.
2. Some specs require the ShopifyAPI be correctly configured, but some specs
   reconfigure it with dummy data. An around(:each) block ensures the original
   configuration is restored after every example is run.

Commit:
Extracts Private Policy Class From ShopifyRefunder

One of the responsibilities taken by ShopifyRefunder is to determine whether
the refund can be issued. I've extracted this responsibility into a private
(for now) CanIssueRefundPolicy object, which can subsequently be separately
unit-tested and promoted as a first-class object.

Commit:
Promotes CanIssueRefundPolicy to First-Class Object

Extracting and promoting the policy object allowed for a significant reduction
in the complexity of the ShopifyRefunder, both implementation and specs. The
refunder now has no knowledge of:

- the conditions under which an error should be raised;
- the relationship between the credit-amount and the transaction-amount.

Commit:
Cleans up ShopifyRefunder Spec

- Consolidates `subject` declarations
- Moves MockError definition closer to invokation for clarity
- Other minor readability tweaks

Commit:
Use Keyword Args in ShopifyRefunder Initializer

With so many arguments, it's greatly preferable to avoid argument-order
dependencies by using keyword arguments. More preferable still would be
reducing the complexity of the object...

Commit:
Delinting

- Separates arguments from injected dependencies for clarity
- Consistent use of symbols for doubles
- Decomposes memoized 'options' into separate arguments

Commit:
Improves Expectations on Refund Creation

The example exercising the creation of the refund didn't set any expectations
on the arguments received by the refund interface. Now it does!

Commit:
Improves Spree::Refund Spec

The spec for 'Refund#pos_order_id' has been improved to exercise its knowledge,
rather than merely its interface.

DanielWright commented:
I'd like to avoid having the changes here go any further beyond the PR's original scope. Can you guys give a once-over to my commits and let me know if anything egregious is missing?

(Also, why is Travis not running here?)

thisiscab commented:
LGTM 👍
